arch/arm/src/nrf52/nrf52_gpiote.c

```diff
@@ -61,9 +61,14 @@ struct nrf52_gpiote_callback_s
  * Private Data
  ****************************************************************************/
 
-/* Interrupt handlers attached to each GPIOTE */
+/* Callbacks attached to each GPIOTE channel */
 
-static struct nrf52_gpiote_callback_s g_gpiote_callbacks[GPIOTE_CHANNELS];
+static struct nrf52_gpiote_callback_s g_gpiote_ch_callbacks[GPIOTE_CHANNELS];
+
+/* Callbacks attached to each GPIO pin */
+
+static struct nrf52_gpiote_callback_s
```

Comment:
What if we introduce a new structure for pin events
```
struct nrf52_gpiote_pin_callback_s
{
  xcpt_t     callback;
  FAR void  *arg;
  uint8_t     pin;
  #ifdef CONFIG_NRF52_HAVE_PORT1
  uint8_t     port;
  #endif
}
```
and configure size of g_gpiote_pin_callbacks from Kconfig? 
With this we can save some resources and eliminate the scanning of all pins.

Reply:
I think this will really complicate Kconfig. You will have to know how many pins will require an interrupt, which is not something that is usually done on any other arch. The alternative you suggest means that you have to search for the complete list of callbacks for the matching pin (trades speed for memory). I don't think that the memory cost-saving of not having all pins be configurable with a callback (256 bytes for the whole port) is worth it for the added complexity of having to know before hand the number of pins you will want to set with callbacks.

There's an underlying issue here IMHO: if you look at stm32_gpiosetevent it makes you think you are setting the interrupt for the pin when you can only set interrupts by groups of pins sharing the interrupt (in nRF52 this group is the whole GPIO port). If you look at how this is used in NuttX, it is used as it would effectively let you set interrupts for pins individually, which means that setting the callback for another pin in the same group will unset the first one. 
In this implementation I intended to explicitly support per-pin callbacks since otherwise you are simply setting one ISR for all GPIOs (assuming one port as in nRF52832). I think that this is the way we should handle pin interrupts to add some value over simply setting a callback for the ISR.